File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonCollectionReader.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.base.pool.ConcurrentObjectPool;
import cn.wjybxx.dson.ext.MarkableIterator;
import cn.wjybxx.dson.internal.DsonInternals;
import cn.wjybxx.dson.io.DsonIOException;
import cn.wjybxx.dson.types.*;

import java.util.*;

/**
 * @author wjybxx
 * date - 2023/6/13
 */
public final class DsonCollectionReader extends AbstractDsonReader {

    private String nextName;
    private DsonValue nextValue;

    public DsonCollectionReader(DsonReaderSettings settings, DsonArray<String> dsonArray) {
        super(settings);
        Objects.requireNonNull(dsonArray);

        Context context = newContext(null, DsonContextType.TOP_LEVEL, null);
        context.header = !dsonArray.getHeader().isEmpty() ? dsonArray.getHeader() : null;
        context.arrayIterator.setBaseIterator(dsonArray.iterator());
        setContext(context);
    }

    /**
     * 设置key的迭代顺序
     * 注意：这期间不能触发{@link #peekDsonType()}等可能导致mark的操作，
     * mark操作会导致key缓存到本地，从而使得外部的keyItr无效。
     *
     * @param defValue key不存在时的返回值；可选择{@link DsonNull#UNDEFINE}
     */
    public void setKeyItr(Iterator<String> keyItr, DsonValue defValue) {
        Objects.requireNonNull(keyItr);
        Objects.requireNonNull(defValue);
        Context context = getContext();
        if (context.dsonObject == null) {
            throw DsonIOException.contextError(List.of(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        context.setKeyItr(keyItr, defValue);
    }

    /**
     * 获取当前对象的所有key。
     * 注意：不可修改返回的集合。
     */
    public Set<String> getkeySet() {
        Context context = getContext();
        if (context.dsonObject == null) {
            throw DsonIOException.contextError(List.of(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        return context.dsonObject.keySet();
    }

    @Override
    protected Context getContext() {
        return (Context) super.getContext();
    }

    @Override
    public void close() {
        Context context = getContext();
        setContext(null);
        while (context != null) {
            Context parent = context.getParent();
            contextPool.release(context);
            context = parent;
        }
        nextName = null;
        nextValue = null;
        super.close();
    }

    // region state

    private void pushNextValue(DsonValue nextValue) {
        this.nextValue = Objects.requireNonNull(nextValue);
    }

    private DsonValue popNextValue() {
        DsonValue r = this.nextValue;
        this.nextValue = null;
        return r;
    }

    private void pushNextName(String nextName) {
        this.nextName = Objects.requireNonNull(nextName);
    }

    private String popNextName() {
        String r = this.nextName;
        this.nextName = null;
        return r;
    }

    @Override
    public DsonType readDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        popNextName();
        popNextValue();

        DsonType dsonType;
        if (context.header != null) { // 需要先读取header
            dsonType = DsonType.HEADER;
            pushNextValue(context.header);
            context.header = null;
        } else if (context.contextType.isArrayLike()) {
            DsonValue nextValue = context.nextValue();
            if (nextValue == null) {
                dsonType = DsonType.END_OF_OBJECT;
            } else {
                pushNextValue(nextValue);
                dsonType = nextValue.getDsonType();
            }
        } else {
            Map.Entry<String, DsonValue> nextElement = context.nextElement();
            if (nextElement == null) {
                dsonType = DsonType.END_OF_OBJECT;
            } else {
                pushNextName(nextElement.getKey());
                pushNextValue(nextElement.getValue());
                dsonType = nextElement.getValue().getDsonType();
            }
        }

        this.currentDsonType = dsonType;
        this.currentWireType = WireType.VARINT;
        this.currentName = INVALID_NAME;

        onReadDsonType(context, dsonType);
        return dsonType;
    }

    @Override
    public DsonType peekDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        if (context.header != null) {
            return DsonType.HEADER;
        }
        if (!context.hasNext()) {
            return DsonType.END_OF_OBJECT;
        }
        if (context.contextType.isArrayLike()) {
            context.markItr();
            DsonValue nextValue = context.nextValue();
            context.resetItr();
            return nextValue.getDsonType();
        } else {
            context.markItr();
            Map.Entry<String, DsonValue> nextElement = context.nextElement();
            context.resetItr();
            return nextElement.getValue().getDsonType();
        }
    }

    @Override
    protected void doReadName() {
        currentName = popNextName();
    }

    // endregion

    // region 简单值

    @Override
    protected int doReadInt32() {
        return popNextValue().asInt32(); // as顺带null检查
    }

    @Override
    protected long doReadInt64() {
        return popNextValue().asInt64();
    }

    @Override
    protected float doReadFloat() {
        return popNextValue().asFloat();
    }

    @Override
    protected double doReadDouble() {
        return popNextValue().asDouble();
    }

    @Override
    protected boolean doReadBool() {
        return popNextValue().asBool();
    }

    @Override
    protected String doReadString() {
        return popNextValue().asString();
    }

    @Override
    protected void doReadNull() {
        popNextValue();
    }

    @Override
    protected Binary doReadBinary() {
        return popNextValue().asBinary().deepCopy(); // 需要拷贝
    }

    @Override
    protected ObjectPtr doReadPtr() {
        return popNextValue().asPointer();
    }

    @Override
    protected ObjectLitePtr doReadLitePtr() {
        return popNextValue().asLitePointer();
    }

    @Override
    protected ExtDateTime doReadDateTime() {
        return popNextValue().asDateTime();
    }

    @Override
    protected Timestamp doReadTimestamp() {
        return popNextValue().asTimestamp();
    }

    // endregion

    // region 容器

    @Override
    protected void doReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = newContext(getContext(), contextType, dsonType);
        DsonValue dsonValue = popNextValue();
        if (dsonValue.getDsonType() == DsonType.OBJECT) {
            DsonObject<String> dsonObject = dsonValue.asObject();
            newContext.header = !dsonObject.getHeader().isEmpty() ? dsonObject.getHeader() : null;
            newContext.dsonObject = dsonObject;
            newContext.objectIterator.setBaseIterator(dsonObject.entrySet().iterator());
        } else if (dsonValue.getDsonType() == DsonType.ARRAY) {
            DsonArray<String> dsonArray = dsonValue.asArray();
            newContext.header = !dsonArray.getHeader().isEmpty() ? dsonArray.getHeader() : null;
            newContext.arrayIterator.setBaseIterator(dsonArray.iterator());
        } else {
            // header
            DsonHeader<String> header = dsonValue.asHeader();
            newContext.dsonObject = header;
            newContext.objectIterator.setBaseIterator(header.entrySet().iterator());
        }
        newContext.name = currentName;

        this.recursionDepth++;
        setContext(newContext);
    }

    @Override
    protected void doReadEndContainer() {
        Context context = getContext();

        // 恢复上下文
        recoverDsonType(context);
        this.recursionDepth--;
        setContext(context.parent);
        returnContext(context);
    }

    // endregion

    // region 特殊接口

    @Override
    protected void doSkipName() {
        popNextName();
    }

    @Override
    protected void doSkipValue() {
        popNextValue();
    }

    @Override
    protected void doSkipToEndOfObject() {
        Context context = getContext();
        context.header = null;
        if (context.dsonObject != null) {
            context.objectIterator.close();
            context.objectIterator.setBaseIterator(Collections.emptyIterator());
        } else {
            context.arrayIterator.close();
            context.arrayIterator.setBaseIterator(Collections.emptyIterator());
        }
    }

    @Override
    protected byte[] doReadValueAsBytes() {
        throw new UnsupportedOperationException();
    }

    // endregion

    // region context

    private static final ConcurrentObjectPool<Context> contextPool = new ConcurrentObjectPool<>(Context::new, Context::reset,
            DsonInternals.CONTEXT_POOL_SIZE);

    private static Context newContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = contextPool.acquire();
        context.init(parent, contextType, dsonType);
        return context;
    }

    private static void returnContext(Context context) {
        contextPool.release(context);
    }

    protected static class Context extends AbstractDsonReader.Context
            implements Iterator<Map.Entry<String, DsonValue>> {

        /** 如果不为null，则表示需要先读取header */
        private DsonHeader<String> header;
        /** 如果不为null，则表示读取Header和Object上下文 */
        private AbstractDsonObject<String> dsonObject;
        /** 迭代器和Context一起缓存 */
        private final MarkableIterator<Map.Entry<String, DsonValue>> objectIterator = new MarkableIterator<>(Collections.emptyIterator());
        private final MarkableIterator<DsonValue> arrayIterator = new MarkableIterator<>(Collections.emptyIterator());

        /** 按照外部key迭代 -- 避免再封装一层增加开销 */
        private Iterator<String> keyItr;
        private DsonValue defValue;

        public Context() {
        }

        @Override
        public Context getParent() {
            return (Context) parent;
        }

        @Override
        public void reset() {
            super.reset();
            header = null;
            dsonObject = null;
            objectIterator.close();
            arrayIterator.close();
            keyItr = null;
            defValue = null;
        }

        /** 该方法重合了迭代器的hasNext，需要兼容 */
        @Override
        public boolean hasNext() {
            if (keyItr != null) {
                return keyItr.hasNext();
            }
            if (dsonObject != null) {
                return objectIterator.hasNext();
            }
            return arrayIterator.hasNext();
        }

        public void markItr() {
            if (dsonObject != null) {
                objectIterator.mark();
            } else {
                arrayIterator.mark();
            }
        }

        public void resetItr() {
            if (dsonObject != null) {
                objectIterator.reset();
            } else {
                arrayIterator.reset();
            }
        }

        public DsonValue nextValue() {
            return arrayIterator.hasNext() ? arrayIterator.next() : null;
        }

        public Map.Entry<String, DsonValue> nextElement() {
            return objectIterator.hasNext() ? objectIterator.next() : null;
        }

        // key-itr

        public void setKeyItr(Iterator<String> keyItr, DsonValue defValue) {
            if (dsonObject == null) throw new IllegalStateException();
            if (objectIterator.isMarking()) throw new IllegalStateException("reader is in marking state");

            this.keyItr = keyItr;
            this.defValue = defValue;
            objectIterator.close();
            objectIterator.setBaseIterator(this);
        }

        @Override
        public Map.Entry<String, DsonValue> next() {
            String key = keyItr.next();
            DsonValue dsonValue = dsonObject.get(key);
            if (dsonValue == null) {
                return Map.entry(key, defValue);
            } else {
                return Map.entry(key, dsonValue);
            }
        }
    }

    // endregion

}

File: csharp/Wjybxx.Dson.Core/src/DsonCollectionReader.cs
﻿#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.Collections;
using System.Collections.Generic;
using Wjybxx.Commons.Collections;
using Wjybxx.Commons.Pool;
using Wjybxx.Dson.Ext;
using Wjybxx.Dson.Internal;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Types;

namespace Wjybxx.Dson
{
/// <summary>
/// 从<see cref="DsonArray{TK}"/>中读取输入
/// </summary>
/// <typeparam name="TName"></typeparam>
public sealed class DsonCollectionReader<TName> : AbstractDsonReader<TName> where TName : IEquatable<TName>
{
#nullable disable
    private TName _nextName;
    private DsonValue _nextValue;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="settings"></param>
    /// <param name="dsonArray">输入</param>
    public DsonCollectionReader(DsonReaderSettings settings, DsonArray<TName> dsonArray)
        : base(settings) {
        if (dsonArray == null) throw new ArgumentNullException(nameof(dsonArray));

        Context context = NewContext(null, DsonContextType.TopLevel, DsonTypes.INVALID);
        context.header = dsonArray.Header.Count > 0 ? dsonArray.Header : null;
        context.arrayIterator.SetBaseIterator(dsonArray.GetEnumerator());
        SetContext(context);
    }

    /// <summary>
    /// 设置key的迭代顺序。
    /// 注意：这期间不能触发<see cref="PeekDsonType"/>等可能导致mark的操作，
    ///  mark操作会导致key缓存到本地，从而使得外部的keyItr无效。
    /// </summary>
    /// <param name="keyItr">Key的迭代器</param>
    /// <param name="defValue">key不存在时的返回值</param>
    public void SetKeyItr(ISequentialEnumerator<TName> keyItr, DsonValue defValue) {
        if (keyItr == null) throw new ArgumentNullException(nameof(keyItr));
        if (defValue == null) throw new ArgumentNullException(nameof(defValue));
        Context context = GetContext();
        if (context.dsonObject == null) {
            throw DsonIOException.ContextError(CollectionUtil.NewList(DsonContextType.Object, DsonContextType.Header), context.contextType);
        }
        context.SetKeyItr(keyItr, defValue);
    }

    /// <summary>
    /// 获取当前对象的所有key。
    /// 注意：不可修改返回的集合。
    /// </summary>
    /// <returns></returns>
    /// <exception cref="DsonIOException"></exception>
    public ICollection<TName> Keys() {
        Context context = GetContext();
        if (context.dsonObject == null) {
            throw DsonIOException.ContextError(CollectionUtil.NewList(DsonContextType.Object, DsonContextType.Header), context.contextType);
        }
        return context.dsonObject.Keys;
    }

    private new Context GetContext() {
        return (Context)context;
    }

    public override void Dispose() {
        Context context = GetContext();
        SetContext(null);
        while (context != null) {
            Context parent = context.Parent;
            contextPool.Release(context);
            context = parent;
        }
        _nextName = default;
        _nextValue = null;
        base.Dispose();
    }

    #region state

    private void PushNextValue(DsonValue nextValue) {
        if (nextValue == null) throw new ArgumentNullException(nameof(nextValue));
        this._nextValue = nextValue;
    }

    private DsonValue PopNextValue() {
        DsonValue r = this._nextValue;
        this._nextValue = null;
        return r!;
    }

    private void PushNextName(TName nextName) {
        this._nextName = nextName;
    }

    private TName PopNextName() {
        TName r = this._nextName;
        this._nextName = default;
        return r;
    }

    public override DsonType ReadDsonType() {
        Context context = GetContext();
        CheckReadDsonTypeState(context);

        PopNextName();
        PopNextValue();

        DsonType dsonType;
        if (context.header != null) { // 需要先读取header
            dsonType = DsonType.Header;
            PushNextValue(context.header);
            context.header = null;
        } else if (context.contextType.IsArrayLike()) {
            DsonValue nextValue = context.NextValue();
            if (nextValue == null) {
                dsonType = DsonType.EndOfObject;
            } else {
                PushNextValue(nextValue);
                dsonType = nextValue.DsonType;
            }
        } else {
            KeyValuePair<TName, DsonValue>? nextElement = context.NextElement();
            if (!nextElement.HasValue) {
                dsonType = DsonType.EndOfObject;
            } else {
                PushNextName(nextElement.Value.Key);
                PushNextValue(nextElement.Value.Value);
                dsonType = nextElement.Value.Value.DsonType;
            }
        }

        this.currentDsonType = dsonType;
        this.currentWireType = WireType.VarInt;
        this.currentName = default;

        OnReadDsonType(context, dsonType);
        return dsonType;
    }

    public override DsonType PeekDsonType() {
        Context context = this.GetContext();
        CheckReadDsonTypeState(context);

        if (context.header != null) {
            return DsonType.Header;
        }
        if (!context.HasNext()) {
            return DsonType.EndOfObject;
        }
        if (context.contextType.IsArrayLike()) {
            context.MarkItr();
            DsonValue nextValue = context.NextValue();
            context.ResetItr();
            return nextValue!.DsonType;
        } else {
            context.MarkItr();
            KeyValuePair<TName, DsonValue>? nextElement = context.NextElement();
            context.ResetItr();
            return nextElement!.Value.Value.DsonType;
        }
    }

    protected override void DoReadName() {
        currentName = PopNextName();
    }

    #endregion

    #region 简单值

    protected override int DoReadInt32() {
        return PopNextValue().AsInt32();
    }

    protected override long DoReadInt64() {
        return PopNextValue().AsInt64();
    }

    protected override float DoReadFloat() {
        return PopNextValue().AsFloat();
    }

    protected override double DoReadDouble() {
        return PopNextValue().AsDouble();
    }

    protected override bool DoReadBool() {
        return PopNextValue().AsBool();
    }

    protected override string DoReadString() {
        return PopNextValue().AsString();
    }

    protected override void DoReadNull() {
        PopNextValue();
    }

    protected override Binary DoReadBinary() {
        return PopNextValue().AsBinary().DeepCopy(); // 需要拷贝
    }

    protected override ObjectPtr DoReadPtr() {
        return PopNextValue().AsPointer();
    }

    protected override ObjectLitePtr DoReadLitePtr() {
        return PopNextValue().AsLitePointer();
    }

    protected override ExtDateTime DoReadDateTime() {
        return PopNextValue().AsDateTime();
    }

    protected override Timestamp DoReadTimestamp() {
        return PopNextValue().AsTimestamp();
    }

    #endregion

    #region 容器

    protected override void DoReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = NewContext(GetContext(), contextType, dsonType);
        DsonValue dsonValue = PopNextValue();
        if (dsonValue.DsonType == DsonType.Object) {
            DsonObject<TName> dsonObject = dsonValue.AsObject<TName>();
            newContext.header = dsonObject.Header.Count > 0 ? dsonObject.Header : null;
            newContext.dsonObject = dsonObject;
            newContext.objectIterator.SetBaseIterator(dsonObject.GetEnumerator());
        } else if (dsonValue.DsonType == DsonType.Array) {
            DsonArray<TName> dsonArray = dsonValue.AsArray<TName>();
            newContext.header = dsonArray.Header.Count > 0 ? dsonArray.Header : null;
            newContext.arrayIterator.SetBaseIterator(dsonArray.GetEnumerator());
        } else {
            // header
            DsonHeader<TName> header = dsonValue.AsHeader<TName>();
            newContext.dsonObject = header;
            newContext.objectIterator.SetBaseIterator(header.GetEnumerator());
        }
        newContext.name = currentName;

        this.recursionDepth++;
        SetContext(newContext);
    }

    protected override void DoReadEndContainer() {
        Context context = GetContext();

        // 恢复上下文
        RecoverDsonType(context);
        this.recursionDepth--;
        SetContext(context.parent!);
        ReturnContext(context);
    }

    #endregion

    #region 特殊

    protected override void DoSkipName() {
        PopNextName();
    }

    protected override void DoSkipValue() {
        PopNextValue();
    }

    protected override void DoSkipToEndOfObject() {
        Context context = GetContext();
        context.header = null;
        if (context.dsonObject != null) {
            context.objectIterator.Dispose();
            context.objectIterator.SetBaseIterator(ISequentialEnumerator<KeyValuePair<TName, DsonValue>>.Empty);
        } else {
            context.arrayIterator.Dispose();
            context.arrayIterator.SetBaseIterator(ISequentialEnumerator<DsonValue>.Empty);
        }
    }

    protected override byte[] DoReadValueAsBytes() {
        throw new InvalidOperationException("Unsupported operation");
    }

    #endregion

    #region context

    private static readonly ConcurrentObjectPool<Context> contextPool = new ConcurrentObjectPool<Context>(
        () => new Context(), context => context.Reset(),
        DsonInternals.CONTEXT_POOL_SIZE);

    private static Context NewContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = contextPool.Acquire();
        context.Init(parent, contextType, dsonType);
        return context;
    }

    private static void ReturnContext(Context context) {
        contextPool.Release(context);
    }

#pragma warning disable CS0628
    protected new class Context : AbstractDsonReader<TName>.Context, ISequentialEnumerator<KeyValuePair<TName, DsonValue>>
    {
        /** 如果不为null，则表示需要先读取header */
        protected internal DsonHeader<TName> header;
        /** 如果不为null，则表示读取Header和Object上下文 */
        protected internal AbstractDsonObject<TName> dsonObject;
        /** 迭代器和Context一起缓存 */
        protected internal MarkableIterator<KeyValuePair<TName, DsonValue>> objectIterator =
            new(ISequentialEnumerator<KeyValuePair<TName, DsonValue>>.Empty);
        protected internal MarkableIterator<DsonValue> arrayIterator = new(ISequentialEnumerator<DsonValue>.Empty);

        /** 按照外部key迭代 -- 避免再封装一层增加开销 */
        private ISequentialEnumerator<TName> keyItr;
        private DsonValue defValue;

        public Context() {
        }

        public new Context Parent => (Context)parent;

        // 这里重合的迭代器的Reset...好在目前没有问题，不然还是得使用额外的KeyItr
        public override void Reset() {
            base.Reset();
            header = null;
            dsonObject = null;
            objectIterator.Dispose();
            arrayIterator.Dispose();
            keyItr = null;
            defValue = null;
        }

        /** 该方法重合了迭代器的hasNext，需要兼容 */
        public bool HasNext() {
            if (keyItr != null) {
                return keyItr.HasNext();
            }
            if (dsonObject != null) {
                return objectIterator.HasNext();
            }
            return arrayIterator.HasNext();
        }

        public void MarkItr() {
            if (dsonObject != null) {
                objectIterator.Mark();
            } else {
                arrayIterator!.Mark();
            }
        }

        public void ResetItr() {
            if (dsonObject != null) {
                objectIterator.Reset();
            } else {
                arrayIterator!.Reset();
            }
        }

        public DsonValue NextValue() {
            return arrayIterator!.HasNext() ? arrayIterator.Next() : null;
        }

        public KeyValuePair<TName, DsonValue>? NextElement() {
            return objectIterator!.HasNext() ? objectIterator.Next() : null;
        }

        // key-itr
        public void SetKeyItr(ISequentialEnumerator<TName> keyItr, DsonValue defValue) {
            if (dsonObject == null) throw new InvalidOperationException();
            if (objectIterator!.IsMarking) throw new InvalidOperationException("reader is in marking state");

            this.keyItr = keyItr;
            this.defValue = defValue;
            objectIterator.Dispose();
            objectIterator.SetBaseIterator(this);
        }

        public bool MoveNext() {
            return keyItr.MoveNext();
        }

        public KeyValuePair<TName, DsonValue> Current {
            get {
                TName key = keyItr.Current;
                if (dsonObject.TryGetValue(key!, out DsonValue dsonValue)) {
                    return new KeyValuePair<TName, DsonValue>(key, dsonValue);
                } else {
                    return new KeyValuePair<TName, DsonValue>(key, defValue);
                }
            }
        }

        object IEnumerator.Current => Current;

        public void Dispose() {
        }
    }

    #endregion
}
}